assets/src/edit-story/components/panels/elementAlignment.js

```diff
@@ -0,0 +1,366 @@
+          id,
+          x,
+          y,
+          width,
+          height,
+          frameX,
+          frameY,
+          frameWidth,
+          frameHeight,
+        };
+      }),
+    [selectedElements]
+  );
+
+  const isJustifyDisabled = isFill || selectedElements.length < 2;
+  const isDistributionDisabled = isFill || selectedElements.length < 3;
+
+  const handleAlignLeft = () => {
+    onSetProperties((properties) => {
+      const { x, y, width, height, rotationAngle } = properties;
+      let offsetX = 0;
+      if (rotationAngle) {
+        const { width: frameWidth } = calcRotatedObjectPositionAndSize(
+          rotationAngle,
+          x,
+          y,
+          width,
+          height
+        );
+        offsetX = (frameWidth - width) / 2;
+      }
+      return {
+        x: boundRect.startX + offsetX,
+      };
+    });
+  };
+
+  const handleAlignCenter = () => {
+    const centerX = (boundRect.endX + boundRect.startX) / 2;
+    onSetProperties((properties) => {
+      const { width } = properties;
+      return {
+        x: centerX - width / 2,
+      };
+    });
+  };
+
+  const handleAlignRight = () => {
+    onSetProperties((properties) => {
+      const { x, y, width, height, rotationAngle } = properties;
+      let offsetX = 0;
+      if (rotationAngle) {
+        const { width: frameWidth } = calcRotatedObjectPositionAndSize(
+          rotationAngle,
+          x,
+          y,
+          width,
+          height
+        );
+        offsetX = (frameWidth - width) / 2;
+      }
+      return {
+        x: boundRect.endX - width - offsetX,
+      };
+    });
+  };
+
+  const handleAlignTop = () => {
+    onSetProperties((properties) => {
+      const { x, y, width, height, rotationAngle } = properties;
+      let offsetY = 0;
+      if (rotationAngle) {
+        const { height: frameHeight } = calcRotatedObjectPositionAndSize(
+          rotationAngle,
+          x,
+          y,
+          width,
+          height
+        );
+        offsetY = (frameHeight - height) / 2;
+      }
+      return {
+        y: boundRect.startY + offsetY,
+      };
+    });
+  };
+
+  const handleAlignMiddle = () => {
+    const centerY = (boundRect.endY + boundRect.startY) / 2;
+    onSetProperties((properties) => {
+      const { height } = properties;
+      return {
+        y: centerY - height / 2,
+      };
+    });
+  };
+
+  const handleAlignBottom = () => {
+    onSetProperties((properties) => {
+      const { x, y, width, height, rotationAngle } = properties;
+      let offsetY = 0;
+      if (rotationAngle) {
+        const { height: frameHeight } = calcRotatedObjectPositionAndSize(
+          rotationAngle,
+          x,
+          y,
+          width,
+          height
+        );
+        offsetY = (frameHeight - height) / 2;
+      }
+      return {
+        y: boundRect.endY - height - offsetY,
+      };
+    });
+  };
+
+  const handleHorizontalDistribution = () => {
+    updatedSelectedElementsWithFrame.sort(
```

Comment:
TBH it bugs me a bit that we, sort of, violating the `useMemo` contract from above by mutating the array in-place. Maybe it's safer to just copy the array as `const sortedElementsWithFrame = [...updatedELementsWithFrame]; sortedElementsWithFrame.sort(...)`?

Reply:
Makes sense for me. Fixed!